spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -104,6 +108,10 @@
 
 	private final Set<Object> groupIds = ConcurrentHashMap.newKeySet();
 
+	private final ThreadLocal<Tuple2<Object, Message<?>>> deferredSends = new ThreadLocal<>();
+
+	private final ThreadLocal<List<Message<?>>> deferredDiscards = new ThreadLocal<>();
```

Comment:
Would you mind to share some explanation for this `ThreadLocal` trick?
Why simple `if...else` around `sendOutputs(result, message);` to release lock if present as an extra argument for the `completeGroup()` is not enough?

Something similar could be done with the `discardMessage()` as well.

Thanks